File: dotnet/Service.cs
using Microsoft.Extensions.Logging;

namespace api
{
    public interface IService
    {
        User GetUser();
    }

    public class Service : IService
    {
        private readonly IRepository _repository;
        private readonly IEmailSender _emailSender;
        private readonly ILogger<Service> _logger;

        public Service(IRepository repository, IEmailSender emailSender, ILogger<Service> logger)
        {
            _repository = repository;
            _emailSender = emailSender;
            _logger = logger;
        }

        public User GetUser()
        {
            _logger.LogInformation($"service: get-user");
            var user = _repository.GetUser("eduard");
            _emailSender.SendWelcomeEmail(user.Email);
            return user;
        }
    }
}

File: nodejs-di/service.js
class Service {
  constructor(repository, emailSender, logger) {
    this.repository = repository;
    this.emailSender = emailSender;
    this.logger = logger;
  }

  getUser(userName) {
    this.logger.log("service", `get-user ${userName}`);
    let user = this.repository.getUser(userName);
    this.emailSender.sendWelcomeEmail(user.email);

    return user;
  }
}

module.exports = Service;
